docs/source/en/llm_optims.md

```diff
@@ -0,0 +1,271 @@
+
+This guide will show you how to use the optimization techniques available in Transformers to accelerate LLM inference.
+
+> [!TIP]
+> Hugging Face also provides [Text Generation Inference (TGI)](https://hf.co/docs/text-generation-inference), a library dedicated to deploying and serving highly optimized LLMs for inference. It includes more optimization features not included in Transformers such as continuous batching for increasing throughput and tensor parallelism for multi-GPU inference.
+
+## Static kv-cache and torch.compile
+
+During decoding, a LLM computes the key-value (kv) values for each input token and since it is autoregressive, it computes the same kv values each time since the generated output becomes part of the input now. This is not very efficient because you're recomputing the same kv values each time.
+
+To optimize this, you can use a kv-cache to store the past values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses Python code into fast and optimized kernels.
+
+The *static kv-cache* solves this issue by fixing the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.
+
+> [!WARNING]
+> Currently, only [Gemma](./model_doc/gemma) and [Llama](./model_doc/llama2) models support static kv-cache and torch.compile.
+
+For this example, let's load the [Gemma](https://hf.co/google/gemma-2b) model.
+
+```py
+from transformers import AutoTokenizer, AutoModelForCausalLM
+
+tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
+
+model = AutoModelForCausalLM.from_pretrained(
+    "google/gemma-2b", device_map="auto"
+)
+```
+
+There are two ways you can configure the model to use a static kv-cache.
+
+<hfoptions id="static-kv">
+<hfoption id="generation_config">
+
+Access the model's `generation_config` attribute and set the `cache_implementation` to "static".
+
+```py
+model.generation_config.cache_implementation = "static"
+```
+
+Call torch.compile on the model to compile it with the static kv-cache.
+
+```py
+compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
+input_text = "Planck's constant is"
+input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
+
+outputs = compiled_model.generate(**input_ids)
+```
+
+</hfoption>
+<hfoption id="setup_cache">
+
+Access the model's `_set_cache` method and pass it the [`StaticCache`] class. This is a more flexible method because it allows you to configure parameters like the maximum batch size and sequence length.
```

Comment:
I think it would be great to think about changing the method name from `_set_cache` to `set_cache` if it is not supposed to be internal.
Also, I think it would be good to explain what is the value that should be put in the `max_cache_len` (maximum number of tokens if I got it right).